File: frontend/src/types/analyzer.ts
/**
 * TypeScript types for Neonatal Report Analyzer
 */

export interface PatientInfo {
  name?: string;
  age_gender?: string;
  uhid?: string;
  referred_by?: string;
  collected_on?: string;
  received_on?: string;
  reported_on?: string;
}

export interface BiochemicalParam {
  parameter: string;
  result: string;
  method: string;
}

export interface TestResult {
  analyte: string;
  value: number;
  reference_range: string;
  unit: string;
  is_normal?: boolean;
  validation_reason?: string;
}

export interface TestRatio {
  ratio: string;
  value: number;
  reference_range: string;
  is_normal?: boolean;
  validation_reason?: string;
}

export interface Abnormality {
  category: string;
  analyte: string;
  value: number;
  reference_range: string;
  reason: string;
  unit: string;
}

export interface AnalysisSummary {
  total_tests: number;
  normal_count: number;
  abnormal_count: number;
  status: 'normal' | 'abnormal';
}

export interface AnalysisResult {
  file_name: string;
  patient_info: PatientInfo;
  biochemical_params: BiochemicalParam[];
  amino_acids: TestResult[];
  amino_acid_ratios: TestRatio[];
  acylcarnitines: TestResult[];
  acylcarnitine_ratios: TestRatio[];
  abnormalities: Abnormality[];
  summary: AnalysisSummary;
}

export interface SingleAnalysisResponse {
  success: boolean;
  message: string;
  result?: AnalysisResult;
  error?: string;
}

export interface FailedReport {
  path: string;
  error: string;
}

export interface BatchAnalysisResponse {
  success: boolean;
  message: string;
  total: number;
  successful: number;
  failed: number;
  normal: number;
  abnormal: number;
  normal_reports: AnalysisResult[];
  abnormal_reports: AnalysisResult[];
  failed_reports: FailedReport[];
}

File: backend/app/schema/analyzer.py
"""
Pydantic schemas for Neonatal Report Analyzer API
"""

from pydantic import BaseModel
from typing import List, Dict, Optional


class PatientInfo(BaseModel):
    """Patient demographic information from report."""
    name: Optional[str] = None
    age_gender: Optional[str] = None
    uhid: Optional[str] = None
    referred_by: Optional[str] = None
    collected_on: Optional[str] = None
    received_on: Optional[str] = None
    reported_on: Optional[str] = None


class BiochemicalParam(BaseModel):
    """Biochemical parameter result."""
    parameter: str
    result: str
    method: str


class TestResult(BaseModel):
    """Individual test result with validation."""
    analyte: str
    value: float
    reference_range: str
    unit: str
    is_normal: Optional[bool] = None
    validation_reason: Optional[str] = None


class TestRatio(BaseModel):
    """Test ratio result with validation."""
    ratio: str
    value: float
    reference_range: str
    is_normal: Optional[bool] = None
    validation_reason: Optional[str] = None


class Abnormality(BaseModel):
    """Abnormal test result details."""
    category: str
    analyte: str
    value: float
    reference_range: str
    reason: str
    unit: str


class AnalysisSummary(BaseModel):
    """Summary statistics of analysis."""
    total_tests: int
    normal_count: int
    abnormal_count: int
    status: str  # 'normal' or 'abnormal'


class AnalysisResult(BaseModel):
    """Complete analysis result for a single PDF."""
    file_name: str
    patient_info: PatientInfo
    biochemical_params: List[BiochemicalParam]
    amino_acids: List[TestResult]
    amino_acid_ratios: List[TestRatio]
    acylcarnitines: List[TestResult]
    acylcarnitine_ratios: List[TestRatio]
    abnormalities: List[Abnormality]
    summary: AnalysisSummary


class SingleAnalysisResponse(BaseModel):
    """API response for single PDF analysis."""
    success: bool
    message: str
    result: Optional[AnalysisResult] = None
    error: Optional[str] = None


class FailedReport(BaseModel):
    """Failed report processing details."""
    path: str
    error: str


class BatchAnalysisResponse(BaseModel):
    """API response for batch PDF analysis."""
    success: bool
    message: str
    total: int
    successful: int
    failed: int
    normal: int
    abnormal: int
    normal_reports: List[AnalysisResult]
    abnormal_reports: List[AnalysisResult]
    failed_reports: List[FailedReport]
